### DISCUSSION TITLE
[Question] robust geometric predicates, polygon triangulation

### DISCUSSION THREAD

**Discussion by artem-ogre** [elalish/manifold#799] [Ideas]
Hello, Emmett (@elalish),

I have recently discovered `manifold` and spent some time reading the documentation and the blog. I very much enjoyed the read and excited about the library. I hope to get to try it hands on in the nearest future. Thank you! 

When reading the part about ["Polygon triangulation"](https://github.com/elalish/manifold/wiki/Manifold-Library#polygon-triangulation) I kept asking myself: could the degeneracy problems you are describing be solved by using robust adaptive floating-point geometric predicates? I can imagine how robust predicates can be useful for solving a handful of challenges faced when implementing robust Boolean operations on 3D meshes: not only for re-triangulating intersecting triangles but also for _exactly_ detecting triangle-triangle intersections with all the possible edge cases.

Did you consider such predicates as part of the related work?

Sorry if this question is already answered by you somewhere. I did my best to search for an answer. I searched GitHub documentation, issues, discussions. I also searched in Julian M. Smith's dissertation and found that it only cites Schewchuk's "Lecture notes on geometric robustness" but in the context not directly related to the predicates.

P.S. I was hesitating whether this should be placed in Discussions, I hope it is okay to open this question as an Issue.

P.P.S
There are several ways of doing robust adaptive precision floating-point geometric predicates, but the most cited paper is probably [this one](https://people.eecs.berkeley.edu/~jrs/papers/robust-predicates.pdf).

**Comment by elalish** [comment]
Thank you! It's certainly a possibility, and one which I have not explored much. The main reason is that exact arithmetic (I'd consider robust geometry predicates to be a form of said) is how most geometry libraries have been made thus far, and I've been underwhelmed by their robustness. I was introduced to computational geometry by Julian Smith's paper, which made me more inclined to look for inexact solutions instead. I also heard anecdotally that exact arithmetic has a serious performance penalty. 

The most fundamental reason I think exact geometry will be problematic is that the internal representations are not exact, nor are the input or output representations (all floating-point). This is why I use the concept of &epsilon;-valid, where we should gracefully handle input that is not quite geometrically valid as though it is perturbed enough to be valid. I think using exact geometry would make this more difficult.

I don't really feel like writing a third triangulator, but if you want to contribute one and show how it's an improvement in robustness and performance, I'd be happy to review.

**Reply by artem-ogre** [reply]
@elalish 
Thanks for taking your time to reply!
> I also heard anecdotally that exact arithmetic has a serious performance penalty.

Yes, in general case this is true, but it's a broad brush: the predicates are designed in a clever way. They are _adaptive_.
The error is tracked and as long it can not change the sign of the predicate no extended precision is used. Only when necessary the precision is extended and an error estimate is updated. Only when necessary the precision is expanded again to the level where exact outcome is guaranteed for any possible floating point inputs. This means that in the vast majority of the invocations the predicates are almost as fast as the inexact ones with only some overhead for error estimation.

> The most fundamental reason I think exact geometry will be problematic is that the internal representations are not exact, nor are the input or output representations (all floating-point).

Right, I see :) Indeed this can be a blocker: constrained Delaunay triangulation with robust predicates can only guarantee correct output for inputs that satisfy two preconditions:
- **no duplicate vertices (exact bit-to-bit duplicates)**: it is possible to work around this by detecting and removing duplicates and remapping the edges
- **no intersecting edges**: there are no ways to work around this in a way that keeps strong mathematical guarantee of a correct output (except maybe perturbing the inputs until intersections go away)

> This is why I use the concept of ε-valid, where we should gracefully handle input that is not quite geometrically valid as though it is perturbed enough to be valid. 

I am very interested what mechanism do you use to achieve that. Do you have any pointers, where should I look for more details? Does it provide a guaranteed correct output for the ε-valid input, or are there still some cases when it may break? Does it involve attempts to perturbate the input?

> I don't really feel like writing a third triangulator, but if you want to contribute one and show how it's an improvement in robustness and performance, I'd be happy to review.

I am already an author of one 😄 ([artem-ogre/CDT](https://github.com/artem-ogre/CDT))
It has been used as a part of mesh Booleans libraries, but given the points above I'm not sure it is the right tool for the job in this case. Regardless, I would love to know more about the domain, constraints, and possible triangulation inputs in the case of `manifold`. Can using exact predicates on inexact inputs can bring any value?🤔 It's an interesting topic to puzzle over for sure.

**Reply by elalish** [reply]
Excellent, I'm happy to meet another who has struggled through making a robust triangulator! As for what mechanism, I take the approach of for each geometric check I make, I test if it's certain (beyond the precision distance that keeps track of my worst-case error), or not. If it's not certain, then I walk the polygon making more checks until I find certainty. 

And of course I need to gracefully handle polygons that are certainly invalid as well, so I need to operate in a way that will always create a manifold result, even if it's not a proper CCW triangulation. My first attempt was a sweep line algorithm based on monotone subdivision, but sweep lines are pretty disastrous with inexact input. Now I have an ear clipping routine that is quite robust because it sorts things by certainty, which helps both with robustness and with triangle quality. Of course the hard part is connecting holes to outer contours, but it does quite well.

My current guarantee is the result is definitely manifold (according to the directed edges input) for absolutely all input. Beyond that, I'd say it's best effort - I'm not sure how I'd go about guaranteeing it produces an &epsilon;-valid triangulation for all &epsilon;-valid polygons. However, I do have a Boolean that conveniently produces an enormous number of very difficult degenerate polygons, and in debug mode it automatically checks every triangulation for validity and creates a test case for any that fail. So, I now have a pretty solid corpus of interesting polygon test cases that you can find [here](https://github.com/elalish/manifold/blob/master/test/polygon_corpus.cpp).

**Comment by pca006132** [comment]
Not sure if I understand this correctly, but it seems that those robust predicates just eliminate errors when checking those predicates, and things like interpolation, matrix multiplication used in union or transformation can still introduce errors. When there is such error, polygons may be self-intersecting and cannot be triangularized even if there is some way to use the robust predicates?

**Reply by artem-ogre** [reply]
Yes, I understand it the same way. If input data are the truth the predicates give truthful answers about the data. For example, in the case of 3D mesh Booleans: predicates tell _exactly_ what triangles intersect and how (including if it's a touching vertex, overlapping edge, etc.). But as soon as you start constructing new geometry (resolving intersections in the case of mesh Booleans): the constructions are inexact. They are a subject to rounding errors. Often correct answers can't be physically represented with IEEE 754 floats. So when invoking predicates on constructed inputs the answers can't be considered truthful anymore.

side note: CGAL calls it "exact predicates, inexact constructions"

**Comment by pca006132** [comment]
I'm thinking about this today, wondering if we can do the following:

1. Use typical Delaunay triangulation with our points.
2. Add polygon edges as constraints. But instead of checking for intersections exactly, we ~~skip if there exists an epsilon perturbation to avoid the collision.~~ keep triangles where the intersecting edge is another constraint edge?
3. For every edge on the polygon, we decide which side to retain by checking edge orientation unless it is hard to decide, i.e., if both sides are not absolutely clockwise based on signed area. Also remove those adjacent triangles from this vertex that are on the removed side.
4. For the hard-to-decide triangles, we look at their neighbors after deciding everything else. If the edge orientation of the kept triangle is wrong, we flip it.

Will try to code this up and see what we can get from this.

**Reply by elalish** [reply]
It's been a long time since I contemplated making a modified Delaunay triangulator, so I don't entirely remember why I never did. What happens to a constrained Delaunay algorithm if two of the constraint edges intersect? Does it crash? I would start by seeing if we can get a Delaunay-type algorithm to give us a manifold result (no concern about CCW triangles) with a simple, but obviously invalid polygon, without adding any new verts. If we can get that to work, then I'd have confidence we can get it to make $\epsilon$-valid output for $\epsilon$-valid input.

**Reply by pca006132** [reply]
> What happens to a constrained Delaunay algorithm if two of the constraint edges intersect? Does it crash?

I think libraries such as CDT require no intersection between constraint edges. Not sure what will happen if there are collisions. And I think this is one of the main problems for us, because invalid polygons require self-intersecting triangulations.

This is why I was thinking about an approximate check when adding constraint edges, which allows slightly intersecting triangles to pass.

**Reply by pca006132** [reply]
Forgot to mention, we will need to deduplicate points for Delaunay, and recover that back by creating degenerate triangles.

**Reply by elalish** [reply]
Yeah, we just have to remember that the manifoldness guarantee is most critical (even with completely invalid geometry). If we can pass that, then we can work on geometric correctness for marginal inputs. I'm just not sure that Delaunay can even do anything at all in e.g. the case of two overlapping squares. It would have to somehow skip entire verts - maybe have a layered set of sweep lines?

**Reply by pca006132** [reply]
Yes two overlapping squares seems to be problematic:

![init](https://github.com/user-attachments/assets/ec7d4a1b-60d6-4ffa-8a6a-1c976e4ba601)

The first step is Delaunay triangulation:

![delaunay](https://github.com/user-attachments/assets/f56f2d6d-7b59-41af-bb09-32583e1f2d07)

The second step is adding constraints. If the intersecting edge is another constraint, we skip that. It will be interesting to look into more complicated cases where only some intersecting edges are constraints. Note that the result here contains overlapping triangles.

![delaunay-constrained](https://github.com/user-attachments/assets/4470e9b6-ffee-4bd7-939c-e25e8355f559)

And then we try to classify the triangles and throw away those that are on the outside. The idea is that for triangles that involve the constraints (2 per edge), we flip the triangle such that the constraints are in the correct direction, and look at the signed area.

![delaunay-constrained-final](https://github.com/user-attachments/assets/1af2c7f0-b00d-468f-a63e-895a71922406)

But this is not topologically correct: the small diagonal in the middle has 4 adjacent triangles. Idk if there is any simple way to fix this. Will this always correspond to self-intersections? Can we somehow flip the endpoints (conceptually) and remove the degenerate?

**Reply by pca006132** [reply]
For the last issue, I guess we can always be certain that such an edge with more than 2 adjacent triangles always corresponds to some self-intersecting edge. This is caused by overlapping triangles, and overlapping triangles can only happen when things self-intersect: when we add the constraints, we allow self-intersection to occur, and there should be no overlapping triangles from Delaunay triangulation otherwise.

I guess I should code things up and see... Maybe with some very brute force approach, and look at the results.

@elalish What do you think?

**Reply by pca006132** [reply]
Collinear points in the middle of edges are common. One reduction rule may be to split a degenerate triangle, which can split the long edge into shorter edges and avoid the collinearity issue.

**Reply by elalish** [reply]
I don't think there's any manifold solution where edges connect the two squares. So, I think delaunay would have to be changed internally to avoid connecting to verts with backward constraint edges. No idea if that's possible, but I'm dubious we can just fix up the manifold after the fact.

**Reply by pca006132** [reply]
I'm thinking about it. If we keep intersecting triangles containing other constraint edges as step 2 above, it breaks manifoldness and breaks the halfedge data structure, in addition to probably something that can't just be fixed later.

Maybe we will need to do some classification/vert swap earlier in this case to select which face to keep? I am trying to get some more problematic cases and play with them. I think I have some idea after thinking about more exotic cases, e.g. square inside square where the inner square is also CW but error caused it to appear as CCW.

**Reply by pca006132** [reply]
Thinking about it, flipping the vertices kind of corresponds to folding it. Consider the following, where the dotted line is the self-intersecting constraint that we want to add, and the red line is the one we want to flip to resolve the constraint.

![image](https://github.com/user-attachments/assets/4f376027-51d6-46a7-a2a2-adac906ffe4d)

The blue triangles are folded:

![image](https://github.com/user-attachments/assets/79e9250b-8e7e-49b6-ad17-24a52f66343c)

So this is still 2-manifold. There are some issues here:

1. Will conflicting constraints always have an edge between them? What should we do if there is none?
2. How will this interact with other constraints? Can we ignore constraints that are not in the folded region, say?
3. There are multiple different possible flips. Which one should we choose? Minimizing the folded area?

**Reply by pca006132** [reply]
> Will conflicting constraints always have an edge between them? What should we do if there is none?

Not necessarily.

![image](https://github.com/user-attachments/assets/4cbd0306-f559-4ec4-b932-fd5f367f148b)

If the longest vertical edge is the conflicting constraint which we add later, it will cross 4 other constraint edges:

![image](https://github.com/user-attachments/assets/14ede770-cb4f-44e3-b345-ea8ea5cf96d8)

**Reply by artem-ogre** [reply]
> > What happens to a constrained Delaunay algorithm if two of the constraint edges intersect? Does it crash?
> 
> I think libraries such as CDT require no intersection between constraint edges. Not sure what will happen if there are collisions. And I think this is one of the main problems for us, because invalid polygons require self-intersecting triangulations.
> 
> This is why I was thinking about an approximate check when adding constraint edges, which allows slightly intersecting triangles to pass.

@elalish @pca006132 
CDT will raise en error by default. It is possible to resolve the intersection by adding a point, but then no guarantee of correctness can be given (because of floating point rounding errors). To my knowledge having robust algorithms that use constructions (e.g., new points at intersections) and floats is hard/unsolved problem.

One option would be to re-triangulate pseudo-polygons but keep old triangulation to preserve both intersecting constraints. But this will not always keep triangulation manifold. Maybe with some care it can be done.

I think _manifold solution can not exist for every configuration with intersecting constraints_. What if my input is three triangles sharing an edge with all 7 edges being constrained? How does one solve this?
![IMG_1110](https://github.com/user-attachments/assets/2eb96969-6820-488a-b20d-c4b2af238ce8)

**Reply by artem-ogre** [reply]
Also allowing intersecting edges will break hole detection algorithms based on neighbor traversal (similar to what @Kimbatt described in the other thread)

**Reply by artem-ogre** [reply]
One more thought: if you don’t care about correctness guarantees: resolving intersections may give you what you need: manifold triangulation where intersections are localized in tiny areas (usually caused by float rounding errors for bad inputs). If you have some test inputs, you can easily try and plug them to CDT. Let me know if I can help.

**Reply by pca006132** [reply]
1. We do not accept adding new points. That breaks our topology.
2. It is not like we don't care about correctness, it is just that correctness is defined differently in this case. We already know that for our inputs it is often impossible to have a correct triangulation by the usual standard, i.e. topologically correct + all ccw and no overlapping triangles. What we want is always topologically correct, and minimize the geometrically invalid regions.
3. Your example is not a valid input for us, it can only be valid as a deduplicated version of some input. That said, I do need to think about this, at least I currently wanted to deduplicate the points and somehow introduce degenerate triangles to fix the topology.
4. Yes I thought about implication on neighbor traversal, but I think that should not be too hard to fix if we know the geometrically invalid region should be small. Even if it requires additional fixup with higher time complexity, such as quadratic on the number of folded triangles, it will be fine because we expect not that many folded ones.

**Reply by elalish** [reply]
Keep in mind, this is why I developed our robust ear-clipping triangulator. It has the excellent guarantee of always producing a manifold result no matter how bad the input geometry is (so, yes, it is possible). I'm not sure if you're hoping to find a faster algorithm or one that can guarantee correct handling of $\epsilon$-valid geometry, or both, but I think refining our current one may bear more fruit than starting over.

**Reply by pca006132** [reply]
Probably. I am just suddenly interested in thinking about if it is possible to get a faster algorithm.

**Comment by starseeker** [comment]
I don't know if it's helpful in this context, but I thought I'd mention this:  https://github.com/Kimbatt/detria

I ended up trying it a while back as a replacement for poly2tri in BRL-CAD (Delaunay triangulation with Steiner points is how we get shading triangles for NURBS surfaces) and it succeeds in producing output for some cases where our (older copy of) poly2tri fork crashed.  I don't know the details, but it mentions using the Shewchuk predicates.

**Reply by elalish** [reply]
Thanks, that does look like a nice library - I'm sure Delaunay will come in handy. I'd be pretty surprised if it handled manifoldness with $\epsilon$-valid inputs, but it might still be a nice place to start.

**Reply by pca006132** [reply]
I think for us, we don't really need or want the typical robust predicate for orient, or exact floating point intersection for the constraints, mainly because we have quite a different correctness criterion.

**Reply by elalish** [reply]
True, though i would bet we could easily replace their colinear / in-circle functions with our own, thus not having to write our own delaunay.

**Reply by starseeker** [reply]
The dev of detria was very responsive when I worked with it previously, so if there are mods they could make that would make it more advantageous for this use case it may be worth asking/having the discussion.

**Reply by pca006132** [reply]
@Kimbatt are you interested in this? Our [wiki](https://github.com/elalish/manifold/wiki/Manifold-Library#overlaps) contains information on why (small) self-intersections will occur in our case and why we cannot allow the triangulator to fix the intersections for us.

We want to see if it is possible to adapt a CDT algorithm to work with the ε-valid polygons we have. We had some discussions above, and the latest one is what I thought of yesterday night. I am not a geometry guy, it may be non-sense :P

And here is an example ε-valid polygon by @elalish to give you some idea why a typical CDT algorithm cannot work:

<img width="953" alt="epsilon-valid" src="https://github.com/user-attachments/assets/6f33ee80-51ca-4183-884a-668871485d17" />

The left is the main polygon with two holes. The middle one is the center of the main polygon, containing self-intersections. The right is the ε-perturbation that removes the self-intersections, showing that the main polygon is indeed ε-valid.
Note that we don't want to compute the ε-perturbation.

For topology reasons, we cannot add vertices (such as intersection vertices), and cannot change the direction of the edges. If there is an edge `u -> v` in the polygon, the triangle containing `u, v` must also go from `u -> v`, even when this makes the triangle go clockwise. Each edge in the triangulation should correspond to exactly 2 triangles.

Due to potential self-intersections in the input and the topology constraints, there will be triangles that are wound clockwise or overlapping. This is fine: our goal is to minimize the geometrically invalid region, or at least small enough, but not necessarily optimal. For this, you can have a look at the [thread above](https://github.com/elalish/manifold/discussions/799#discussioncomment-13340835).

**Reply by Kimbatt** [reply]
I don't think detria would work well for this specific problem, since it was kinda designed from the ground-up to have exact arithmetic only. Having epsilons in the code would break many assumptions, and probably would require a totally different approach to handle these kinds of polygons.

It'd probably be simple to replace the orientation and incircle functions with non-exact ones, but the code would be likely to break in some other place.

I'll think about this though, and get back to you if I come up with something.

In the meantime, here's a high-level overview of the algorithm used in detria, this might give you some ideas for implementing something that works for your case.

---

The algorithm has 3 main phases:
- Initial triangulation: just a valid triangulation that connects all vertices (can be delaunay). This usually includes additional triangles that are only part of the convex hull (and is "useless" for the final result), but these will be generated as a side-effect.
- Create constrained edges: for each constrained edge, ensure that an edge exists between the two vertices. The current algorithm fails here if there are intersecting constrained edges.
- Classify triangles: for each triangle, determine the location of the triangle - can be either "interior", "hole", or "convex hull"

For example, if we want to triangulate a capital E shape:
![image](https://github.com/user-attachments/assets/647df4dc-2c27-42e0-ab61-ac8ca8622fbd)
First, sort all points by their x coordinate, and if there are any vertices with the same x coordinate, then sort those by their y coordinate.
If there are any duplicate points (these will be next to each other after sorting), then the triangulation fails, because this would create degenerate triangles.
Then, create a triangle between the first three vertices (note that this is not always possible, for example, they can be collinear).
![image](https://github.com/user-attachments/assets/e5fcbcc8-d414-4ec5-b041-a95d895d8b7b)

Finally, for each of the remaining vertices (in the sorted order), add triangles from that vertex to the triangulation:
![image](https://github.com/user-attachments/assets/1c5fa97f-ee72-4960-8d3b-822c58052551)
Since the vertices are sorted, there will always be at least one "visible" edge, so we can always create at least one triangle.
For delaunay triangulations, there is an additional "edge flip" step, which checks and flips all affected edges that are non-delaunay.
I didn't do a delaunay triangulation in this example, because it happens to create all the required constrained edges by default, so the "create constrained edges" step would be empty.

In the second phase, go through all constrained edges (vertex pairs), and for each vertex pair, delete edges that are between these vertices. Then add a single edge that connects the vertices.
Finally, re-triangulate the two sides of the edge. This "smaller" triangulation is similar to a monotone triangulation, and can always be done in linear time.
![image](https://github.com/user-attachments/assets/a0852ddd-5bf5-420d-8400-d5526ec54587)
It's possible that an edge already exists between a constrained edge's vertices, so in that case, there's nothing to do.

In the third phase, the triangles are classified by their location. We start from a single triangle for which we know the location (e.g. we pick an edge from the convex hull, and if it's not a constrained edge, then we know it's a "convex hull" triangle)
Then we visit the neighbor triangles, and we keep track of if we are inside/outside. Stepping over a constrained edge (that is part of an outline or hole) flips the "insideness" value.
![image](https://github.com/user-attachments/assets/25270b31-b58f-422d-9ee8-624ee49336dd)

Note that there are some edge cases throughout the entire triangulation process that require extra attention, like a point exactly on a constrained edge (which would create degenerate triangles, or would introduce a topology change), a hole inside another hole, etc.

Hope this is useful for something, or at least help you to get some ideas :D
